### ISSUE TITLE
FPS Tutorial: Bullet material causes pause on first instancing

### ISSUE THREAD

**Issue by ntfwc** [godotengine/godot-docs#1670]
On my system, when you first instance a bullet from the pistol, there is a ~1 second pause. Afterwards there is no problem. It confused me at first, because the bullet scene is preloaded, so instancing it the first time shouldn't cause so much delay. So I suspected it had to do with shader compiling.

I figured out that the bullet material "GunMaterial.material" was causing the delay. It has transparency enabled and the depth draw mode is "Opaque Pre-Pass". So, since it is also being used elsewhere, I copied it to a separate material for the bullet, disabled transparency, and set the depth draw mode to "Opaque Only", and this fixed the pausing. Note: I just wasn't sure if other things using the material needed the transparency, so maybe it didn't need to be copied.

Reverting to the old material, I tried it with the shader cache enabled. After a warmup run, the initial pause was much smaller, losing about 1 frame, but it is still noticeable if you are moving.

The easy solution here is just to change the material settings, but I wonder if there is a better way to handle this. Ideally there would be a way to preload the shaders for things without having to show anything to the user.

About my setup:
Graphics type: Integrated
Graphics chip: Intel HD Graphics 5500
Graphics driver: Mesa 18.0.5
Godot version: 3.0.5 64-bit X11
OS: Xubuntu 16.04

**Comment by TwistedTwigleg** [comment]
Good to know what causes the slight delay! I noticed the delay, but I never thought to look at the material.

As for solutions, I'm not sure what the best thing to do is. Personally I would just change the material settings and/or make the bullet use it's own material, but in theory the material should not be causing a issue since it's preloaded, right? Maybe it is a engine bug?

**Comment by ntfwc** [comment]
So a workaround I tried that works is to setup a sort of "zoo" frame. You set it up so instances of the spawnable meshes are shown and destroyed after the first frame. It seems that it can be hidden behind something, but it has to be in the line of sight of the camera to work.

This is obviously overkill for the tutorial, so yeah, I'd say changing the material is the way to go there.

It does seem that there is an open engine bug related to this: [#13954](https://github.com/godotengine/godot/issues/13954)

**Comment by CoolOppo** [comment]
> I tried it with the shader cache enabled

What shader cache? Are you talking about in your driver settings?

Also, this should be closed, since it's really just a duplicate of #13954, as you already acknowledged.

**Comment by ntfwc** [comment]
> > I tried it with the shader cache enabled
> 
> What shader cache? Are you talking about in your driver settings?

Yeah, I was talking about the the driver shader cache

**Comment by TwistedTwigleg** [comment]
I finally got around to changing the material and disabling transparency. Now the delay when firing the initial bullet is much, much more reduced.

Many thanks for letting me know about it @ntfwc, and my apologizes for the massive delay in actually implementing the change! Things got busy and I forgot.

I also accidental forgot that you, @ntfwc, found the solution! I would have included you in the git log if I remembered! I looked through the issues on the FPS Github page and couldn't find the issue. I only thought to look at the documentation *after* pulling the PR. My apologizes!

With the material changed, I think we can close this issue.

The FPS tutorial will need to be updated with the recent changes on the Github repository, which I'll try to get to this weekend or whenever I can make the time.
